Makefile

```diff
@@ -1,137 +1,205 @@
+PHP_CS_FIXER_CACHE=build/cache/.php_cs.cache
+
+PHPSTAN=./.tools/phpstan
 PHPSTAN_URL="https://github.com/phpstan/phpstan/releases/download/0.11.5/phpstan.phar"
 
-FLOCK=./devTools/flock
+PHPUNIT=vendor/bin/phpunit
+
+INFECTION=./build/infection.phar
 
-DOCKER_RUN=docker run -t --rm -v "$$PWD":/opt/infection -w /opt/infection
+DOCKER_RUN=docker run --tty --rm --volume "$$PWD":/opt/infection --workdir /opt/infection
 DOCKER_RUN_72=$(FLOCK) devTools/*php72*.json $(DOCKER_RUN) infection_php72
+DOCKER_RUN_72_IMAGE=devTools/Dockerfile-php72-xdebug.json
 DOCKER_RUN_73=$(FLOCK) devTools/*php73*.json $(DOCKER_RUN) infection_php73
+DOCKER_RUN_73_IMAGE=devTools/Dockerfile-php73-xdebug.json
 
-.PHONY: all
-#Run all checks, default when running 'make'
-all: analyze test
+FLOCK=./devTools/flock
 
-#Non phony targets for phars etc.
-vendor: composer.json composer.lock
-	composer install
 
-build/cache:
-	mkdir -p build/cache
+#
+# Commands (phony targets)
+#---------------------------------------------------------------------------
 
-./php-cs-fixer-v2.phar:
-	wget $(PHP-CS-FIXER_URL)
-	chmod a+x ./php-cs-fixer-v2.phar
+.PHONY: compile
+compile:	## Bundles Infection into a PHAR
+compile: $(INFECTION)
 
-./phpstan.phar:
-	wget $(PHPSTAN_URL)
-	chmod a+x ./phpstan.phar
+.PHONY: cs-fix
+cs-fix:	  	## Runs PHP-CS-Fixer
+cs-fix: $(PHP_CS_FIXER)
+	# TODO: disable verbose mode?
+	$(PHP_CS_FIXER) fix -v --cache-file=$(PHP_CS_FIXER_CACHE)
 
-#All tests, (infection itself, phpunit, e2e) for different php version/ environments (xdebug or phpdbg)
-.PHONY: test test-unit test-infection-phpdbg test-e2e-phpdbg test-infection-xdebug test-e2e-xdebug
-test: test-unit test-infection-phpdbg test-e2e-phpdbg test-infection-xdebug test-e2e-xdebug
-	# All tests finished without errors
+.PHONY: cs-check
+cs-check:	## Runs PHP-CS-Fixer in dry mode
+cs-check: $(PHP_CS_FIXER)
+	# TODO: disable verbose mode?
+	$(PHP_CS_FIXER) fix -v --cache-file=$(PHP_CS_FIXER_CACHE) --dry-run --stop-on-violation
 
-.PHONY: test-unit test-unit-72 test-unit-73
-#php unit tests
-test-unit: test-unit-72 test-unit-73
+.PHONY: phpstan
+phpstan:  	## Runs PHPStan
+phpstan: vendor $(PHPSTAN)
+	$(PHPSTAN) analyse src --level=max --configuration ./devTools/phpstan-src.neon --no-interaction --no-progress
+	$(PHPSTAN) analyse tests --level=4 --configuration ./devTools/phpstan-tests.neon --no-interaction --no-progress
 
-test-unit-72: build-xdebug-72
-	$(DOCKER_RUN_72) $(PHPUNIT)
+.PHONY: analyze
+analyze:	## Runs CS fixers, static analyzers and various other checks
+analyze: cs-check analyze-ci
 
-test-unit-73: build-xdebug-73
-	$(DOCKER_RUN_73) $(PHPUNIT)
+.PHONY: analyze-ci
+analyze-ci:	## Runs static analyzers and various other checks
+analyze-ci: phpstan validate
 
-.PHONY: test-infection-phpdbg test-infection-phpdbg-72 test-infection-phpdbg-73
-#infection with phpdbg
-test-infection-phpdbg: test-infection-phpdbg-72 test-infection-phpdbg-73
+.PHONY: auto-review
+auto-review:	## TODO: find a suitable description
```

Comment:
TODO: I quite didn't get what this command was about neither when it's supposed to be run

Reply:
Originally the auto-review tests weren't part of the normal phpunit run, but that got changed.

I'm thinking we can safely remove this make command. 

And on another note, i think we may be able to replace some of the auto-review tests with PHPStan rules.

Reply:
:+1: will remove it then